### ISSUE TITLE
Profile cannot change settings on book close when switching via history

### ISSUE THREAD

**Issue by thialfi17** [koreader/koreader#12865]
* KOReader version: both current git + 2024.11
* Device: Emulator + Kobo

#### Issue
Using the profiles plugin to change some settings on a per book/directory basis I cannot get the setting to change back automatically if I don't quit to the file manager. This is useful for certain books/manga I would like to either toggle the frontlight, warmth or orientation lock on and I have them separated in different folders.

I have tried digging into this to fix it and all of ReaderReady/ClosedDocument events are firing correctly for the Profile and the profile is running correctly. It appears that the events fired in the ClosedDocument callback are never triggering and my current guess is that the widgets responsible for changing these settings are being closed before the "ToggleXXX" event is being handled. I'm not sure why this works when opening the file manager.

I haven't tested with every event but I suspect this will be true for any event in the dispatcher system. I'm not familiar enough with the event and UI systems to create a fix but hopefully this information may help someone else.
 
#### Steps to reproduce
Create two profiles:
- one that auto-executes on book opening when book path contains XXX. Action can be toggle fullscreen, toggle backlight etc
- one that auto-executes on book closing when book path contains XXX. Action can be toggle fullscreen, toggle backlight etc

Open a book that satisfies the conditions and see setting changed. Exit to file manager and verify setting was correctly changed again.
Open a book that satisfies the conditions and see setting changed. Open history and open another book that doesn't satisfy conditions and see setting wasn't changed when the original book closed.
Open a book that satisfies the conditions and see setting changed. Open history and open another book that satisfies the conditions and see setting was toggled (unset).

##### `crash.log` (if applicable)
No crashes, just incorrect behaviour. Can potentially add some logs later but they were very noisy.

**Comment by Frenzie** [comment]
Added in #12564.

> Can potentially add some logs later but they were very noisy.

It might be useful to add some logging in `function Profiles:onCloseDocument()` to confirm if your hunch is on the right track.

**Comment by hius07** [comment]
TL;DR: the event sent by Dispatcher is handled when closing Reader for FM and is not handled when switching documents.

In details: a profile is triggered by the `CloseDocument` event.

**If we close Reader for FM:**
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/apps/reader/readerui.lua#L827-L829
then `ReaderUI:onClose()` does
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/apps/reader/readerui.lua#L783
The profile is executed, passes the action to Dispatcher and Dispatcher does:
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/dispatcher.lua#L1189
UIManager gets the call and does
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/ui/uimanager.lua#L911
All OK.

**If we switch documents** (open another document from the History):

`ReaderUI:switchDocument()` does the same
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/apps/reader/readerui.lua#L870
`CloseDocument` event is sent, the profile is executed, Dispatcher sends the event to UIManager, but then UImanager does not do correctly:
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/ui/uimanager.lua#L911

Looks like the `top_widget` is not the same in these cases.

@NiLuJe can you please look at it.

**Comment by hius07** [comment]
To reproduce quickly: create a profile with one action: Set night mode on.
Set auto-exec the profile: on document closing - always.
Set night mode off - open a document - close - night mode is on.
Set night mode off - open a document - open another document from History - night mode remains off.

**Comment by NiLuJe** [comment]
> @NiLuJe can you please look at it.

Won't have time for now, ping me again in early January ;).

That said, your analysis looks sound, but I'm not familiar enough with History to say how that'll look like to UIManager without checking.

**Comment by hius07** [comment]
History is closed at the moment, the difference is in ReaderUi: `onHome` or `switchDocument`.

**Comment by hius07** [comment]
Got it.
When we call `onHome()` (i.e. close Reader), the top widget to handle the event fired by the profile, is FM, and it handles the event well.
When we call `switchDocument()`, the top widget is
https://github.com/koreader/koreader/blob/604e5f30191f9a948853ba3a2ccf3f3363e1db5d/frontend/apps/reader/readerui.lua#L621-L625
Definitely, it cannot handle.

**Comment by hius07** [comment]
@thialfi17 a workaround (until/if we fix it) is to check "Ask to execute" the profile.

**Comment by NiLuJe** [comment]
> Got it.

Nice!

Is it `invisible` at the time? (my train of thought being we can probably safely skip over invisible widgets in UIManager, otherwise, this'll take some more thinking ;).

**Comment by hius07** [comment]
It is visible, opening another document while the previous one is still on the screen (switching).
I tried to remove this InfoMessage, but the problem is that at the moment **there are no widgets** in `self._window_stack`, so UIManager cannot call the event handler.

**Comment by NiLuJe** [comment]
Oh, yeaaah, that's going to be a problem ^^.

**Comment by hius07** [comment]
We execute profile in NextTick
https://github.com/koreader/koreader/blob/0386e2619eb5c11367980741af062d3865ea4e8c/plugins/profiles.koplugin/main.lua#L923-L925
so the previous Reader is closed, the next one is not opened yet.

**Comment by hius07** [comment]
> Is switching to `tickAfterNext` enough to deal with this?

Yes, this works.
